### PR TITLE
omicron-package: work around cargo issue 8157

### PR DESCRIPTION
**Pull request by iliana**
#5799 modified the `cargo build` command line omicron-package runs. Previously it built up a list of packages to be built using the `-p` flag; that PR changed it to use `--bin`. The goal was to build only the binaries that are necessary for shipping; this avoids building sled-agent-sim during releng, for instance.

We did not realize it at the time, but this invited the specter of https://github.com/rust-lang/cargo/issues/8157 to wreak havoc; namely:

- Without `--package`, Cargo uses the `default-members` key of the workspace Cargo.toml to determine which packages to build. `--bin` does not cause the same thing to happen; saying `--bin` does _not_ imply `--package [the package that the bin belongs to]`.
- `omicron-dev` belongs to `default-members` and has a normal dependency on `nexus-test-utils`, which enables the `"testing"` feature of `nexus-db-queries`.

https://github.com/oxidecomputer/omicron/issues/7208 is a known result of this problem, but there might be more.

Fortunately the solution seems fairly easy, without reverting the relevant changes from #5799: use _both_ `--package` and `--bin`. With this change, the `"testing"` feature is no longer shown in the `cargo build --unit-graph` and `nm target/release/nexus | demangle | grep validate_volume_invar` no longer shows any matching testing-only symbols.

For posterity, the full command omicron-package runs is now (from the log): `"cargo" "build" "--package" "dns-server" "--package" "internal-dns-cli" "--package" "omicron-clickhouse-admin" "--package" "omicron-cockroach-admin" "--package" "omicron-gateway" "--package" "omicron-nexus" "--package" "omicron-omdb" "--package" "omicron-sled-agent" "--package" "oximeter-collector" "--package" "oxlog" "--package" "sp-sim" "--package" "wicket" "--package" "wicketd" "--package" "zone-setup" "--bin" "clickhouse-admin-keeper" "--bin" "clickhouse-admin-server" "--bin" "clickhouse-admin-single" "--bin" "clickhouse-schema-updater" "--bin" "cockroach-admin" "--bin" "dns-server" "--bin" "dnsadm" "--bin" "dnswait" "--bin" "mgs" "--bin" "nexus" "--bin" "omdb" "--bin" "oximeter" "--bin" "oxlog" "--bin" "schema-updater" "--bin" "sled-agent" "--bin" "sp-sim" "--bin" "wicket" "--bin" "wicketd" "--bin" "zone-bundle" "--bin" "zone-setup" "--features" "switch-asic" "--release"`

### CODE REVIEW

File: package/src/bin/omicron-package.rs [hunk 1]
```diff
@@ -188,6 +192,8 @@ async fn do_for_all_rust_packages(
         // If this is a Rust package...
         if let PackageSource::Local { rust: Some(rust_pkg), .. } = &pkg.source {
             let plan = if rust_pkg.release { &mut release } else { &mut debug };
+            // Add the package name to the plan
```
Comment (sunshowers) [inline package/src/bin/omicron-package.rs:195]: Could you add a comment here about why this is necessary? Just a link to the issue or this PR would do.

Also worth noting somewhere (if not already done) that when `source.rust` is defined, the table key (e.g. the `omicron-nexus` in `[package.omicron-nexus]`) is the name of the corresponding Rust crate.

**Review by sunshowers — APPROVED** [review]
thanks for jumping so quickly on this!

File: package/src/bin/omicron-package.rs (hunk 1 above)
Comment (jmpesp) [reply package/src/bin/omicron-package.rs:195]: It may also be useful to link to https://github.com/rust-lang/cargo/issues/8157 ?

**Review by jmpesp — APPROVED** [review]

File: package/src/bin/omicron-package.rs (hunk 1 above)
Comment (davepacheco) [reply package/src/bin/omicron-package.rs:199]: > Also worth noting somewhere (if not already done) that when source.rust is defined, the table key (e.g. the omicron-nexus in [package.omicron-nexus]) is the name of the corresponding Rust crate.

This is mentioned here:
https://github.com/oxidecomputer/omicron/blob/d01b2ee56a24e3ec36a75c91d7e4959c9bc87c16/package-manifest.toml#L16-L17

but I agree it's non-obvious and it wouldn't hurt to mention it in other places where a reader might not know.

Comment (jmpesp) [reply package/src/bin/omicron-package.rs:199]: To be clear, in the comment here. I know the PR has it!

Comment (iliana) [reply package/src/bin/omicron-package.rs:199]: https://github.com/oxidecomputer/omicron/pull/7218/commits/acbe39a60eea69d20926a68b480fb04b4df03bd3 does most of this; I think the particular detail about how to write package-manifest.toml is documented in the file itself:

https://github.com/oxidecomputer/omicron/blob/d01b2ee56a24e3ec36a75c91d7e4959c9bc87c16/package-manifest.toml#L12-L17

But I think I improved the comments to make it clearer in that code's context.

**Pull request merged**